compiler/src/iree/compiler/Preprocessing/Common/ApplyPDLPatterns.cpp

```diff
@@ -0,0 +1,529 @@
+    processInputType(type);
+  }
+
+  int64_t numTensorOperands = (int64_t)entryPointInputTypes.size();
+  int64_t numScalarOperands = (int64_t)entryPointScalarInputTypes.size();
+  entryPointInputTypes.append(entryPointScalarInputTypes);
+  entryPointInputTypes.append(totalNumDynamicDims, rewriter.getIndexType());
+
+  auto entryPointFnType = FunctionType::get(context, entryPointInputTypes,
+                                            /*results=*/TypeRange{});
+  auto entryPointFn =
+      rewriter.create<func::FuncOp>(loc, entryPointFnName, entryPointFnType);
+  Region &body = entryPointFn.getBody();
+  SmallVector<Location> locs(entryPointInputTypes.size(), loc);
+  rewriter.createBlock(&body, body.begin(), entryPointInputTypes, locs);
+
+  auto entryPointArgs = entryPointFn.getArguments();
+  auto tensorArgs = entryPointArgs.take_front(numTensorOperands);
+  auto scalarArgs = entryPointArgs.slice(numTensorOperands, numScalarOperands);
+  auto dynamicDimArgs = entryPointArgs.take_back(totalNumDynamicDims);
+  SmallVector<Value> callOperands;
+  Value zero = rewriter.create<arith::ConstantIndexOp>(loc, 0);
+
+  // Method to marshall tensor types into call operands.
+  auto marshallTensorTypes = [&](RankedTensorType tensorType) {
+    int64_t numDynamicDims = tensorType.getNumDynamicDims();
+    auto dynamicDims = dynamicDimArgs.take_front(numDynamicDims);
+    auto [basePtr, offset] = getBasePtrAndOffsetForTensor(
+        rewriter, loc, tensorType, tensorArgs.front(), zero, dynamicDims);
+    callOperands.push_back(basePtr);
+    callOperands.push_back(offset);
+    tensorArgs = tensorArgs.drop_front();
+    dynamicDims = dynamicDims.drop_front(numDynamicDims);
+  };
+  // Method to marshall input types into call operands.
+  auto marshallInputTypes = [&](Type type) {
+    if (type.isIntOrFloat()) {
+      callOperands.push_back(scalarArgs.front());
+      scalarArgs = scalarArgs.drop_front();
+      return;
+    }
+    marshallTensorTypes(type.cast<RankedTensorType>());
+  };
+
+  for (auto type : inputTypes) {
+    marshallInputTypes(type);
+  }
+  for (auto type : resultTypes) {
+    marshallTensorTypes(type.cast<RankedTensorType>());
+  }
+  for (auto type : otherOperandTypes) {
+    marshallInputTypes(type);
+  }
+
+  rewriter.create<func::CallOp>(loc, externalFn, callOperands);
+  rewriter.create<func::ReturnOp>(loc, /*operands=*/ValueRange{});
+  return entryPointFn;
+}
+
+// Generate the `hal.executable` that calls into the external function.
+// Return the nested symbol reference to the entry point function generated.
+static SymbolRefAttr
+createStreamExecutableOp(PatternRewriter &rewriter, Operation *rootOp,
+                         StringRef externalFnName, TypeRange inputTypes,
+                         TypeRange resultTypes, TypeRange otherOperandTypes) {
+  auto moduleOp = rootOp->getParentOfType<ModuleOp>();
+  if (!moduleOp) {
+    llvm_unreachable("found op without surrounding module");
+  }
```

Comment:
this should never be possible (there's always a builtin.module wrapping everything)
if you want to assert that for readability then use assert instead of unreachable
if you find it is possible, then this needs to be proper error handling instead of unreachable
```suggestion
```

(as above: never use unreachable or assert for anything that may be triggered by user inputs)